### PR TITLE
Release TimeSformer's code and update some docs

### PR DESCRIPTION
**Pull request by HydrogenSulfate**
1. add TimeSformer's performance in README.md and README_cn.md
2. add TimeSformer's config file in `configs/recognition/timesformer/timesformer_k400_videos.yaml`
3. add TimeSformer's en doc and Zh-CN doc
4. add TimeSformer's two image about main arch and attention arch
5. add **UniformCrop** in `paddlevideo/loader/pipelines/augmentations.py`, and corresponding **UniformCropMetric** in `paddlevideo/metrics/uniform_crop_metric.py`
6. add pyav's video decode method in `paddlevideo/loader/pipelines/decode.py`
7. add linspace_sample strategy in `paddlevideo/loader/pipelines/sample.py`
8. add **VisionTransformer** in `paddlevideo/modeling/backbones/vit.py`
9. update the different forward method in `paddlevideo/modeling/framework/recognizers/recognizer2d.py ` while using TimeSformer and ViT based model
10. delete **coloring** and **AverageMeter** two unused import statement in `paddlevideo/tasks/train.py`
11. add an new training strategy **gradient accumulation** for TimeSformer and follow-up models, if used, in `paddlevideo/tasks/train.py`
12. fixed a bug when record the learning_rate in training process, when batch_size=1.
13. update the load_ckpt function for TimeSformer and follow-up ViT based models
14. update run.sh for TimeSformer
15. add a new InputSpec in `tools/export_model.py` for TimeSformer, even though there still remain some todo for inference model
16. add **TimeSformer_Inference_helper** class in `tools/utils.py `
17. add inference for timesformer and corresponding content in docs
18. fixed some hyper-link bugs in start.md
19. Add #Params, FLOPs and IPS metric in docs

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -35,23 +35,23 @@ PaddleVideo is a toolset for video recognition, action localization, and spatio
...
-| :--------------- | :--------: | :------------: | :------------: | :------------: |
-| action recgonition | [**PP-TSM**](./docs/zh-CN/model_zoo/recognition/pp-tsm.md) | [Kinetics-400](./docs/zh-CN/dataset/k400.md) | Top-1 | **76.16** |
-| action recgonition | [**PP-TSN**](./docs/zh-CN/model_zoo/recognition/pp-tsn.md) | [Kinetics-400](./docs/zh-CN/dataset/k400.md) | Top-1 | **73.68** |
-| action recgonition | [SlowFast](./docs/zh-CN/model_zoo/recognition/slowfast.md) | [Kinetics-400](./docs/zh-CN/dataset/k400.md) | Top-1 | 75.84 |
-| action recgonition | [TSM](./docs/zh-CN/model_zoo/recognition/tsm.md) | [Kinetics-400](./docs/zh-CN/dataset/k400.md) | Top-1 | 71.06 |
-| action recgonition | [TSN](./docs/zh-CN/model_zoo/recognition/tsn.md) | [Kinetics-400](./docs/zh-CN/dataset/k400.md) | Top-1 | 69.81 |
-| action recgonition | [AttentionLSTM](./docs/zh-CN/model_zoo/recognition/attention_lstm.md) | [Youtube-8M](./docs/zh-CN/dataset/youtube8m.md) | Hit@1 | 89.0 |
-| action detection| [BMN](./docs/zh-CN/model_zoo/localization/bmn.md) | [ActivityNet](./docs/zh-CN/dataset/ActivityNet.md) |  AUC | 67.23 |
+| Field | Model | Dataset | Metrics | ACC% | #Params | FLOPs | IPS |
```
Comment (huangjun12) [inline README.md:38]: Params | FLOPs | IPS 这三个指标先不放到首页里。
后面放到benchmark下

File: README_cn.md [hunk 2]
```diff
@@ -62,15 +62,16 @@ PaddleVideo飞桨视频模型开发套件，旨在帮助开发者更好的进行
...
-| 视频分类 | [**PP-TSN**](./docs/zh-CN/model_zoo/recognition/pp-tsn.md) | [Kinetics-400](./docs/zh-CN/dataset/k400.md) | Top-1 | **73.68** |
-| 视频分类 | [SlowFast](./docs/zh-CN/model_zoo/recognition/slowfast.md) | [Kinetics-400](./docs/zh-CN/dataset/k400.md) | Top-1 | 75.84 |
-| 视频分类 | [TSM](./docs/zh-CN/model_zoo/recognition/tsm.md) | [Kinetics-400](./docs/zh-CN/dataset/k400.md) | Top-1 | 71.06 |
-| 视频分类 | [TSN](./docs/zh-CN/model_zoo/recognition/tsn.md) | [Kinetics-400](./docs/zh-CN/dataset/k400.md) | Top-1 | 69.81 |
-| 视频分类 | [AttentionLSTM](./docs/zh-CN/model_zoo/recognition/attention_lstm.md) | [Youtube-8M](./docs/zh-CN/dataset/youtube8m.md) | Hit@1 | 89.0 |
-| 视频动作定位| [BMN](./docs/zh-CN/model_zoo/localization/bmn.md) | [ActivityNet](./docs/zh-CN/dataset/ActivityNet.md) |  AUC | 67.23 |
+| 领域               |                             模型                             |                       数据集                       | 精度指标 |   精度%   |   参数量    |  浮点运算量   | 每秒图片处理数 |
+| :----------------- | :----------------------------------------------------------: | :------------------------------------------------: | :------: | :-------: | :---------: | :-----------: | :------------: |
```
Comment (huangjun12) [inline README_cn.md:66]: 同英文对应修改。

File: docs/zh-CN/model_zoo/recognition/timesformer.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+UCF101数据下载及准备请参考[UCF-101数据准备](../../dataset/ucf101.md)
+
+
+## 模型训练
+
+### Kinetics-400数据集训练
+
+#### 下载并添加预训练模型
+
+1. 下载图像蒸馏预训练模型[ViT_base_patch16_224](https://paddle-imagenet-models-name.bj.bcebos.com/dygraph/ViT_base_patch16_224_pretrained.pdparams)作为Backbone初始化参数，或通过wget命令下载
```
Comment (huangjun12) [inline docs/zh-CN/model_zoo/recognition/timesformer.md:39]: TODO: download and save

File: docs/zh-CN/model_zoo/recognition/timesformer.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    ```
+    
+- 开启amp混合精度训练，可加速训练过程，其训练启动命令如下：
+
+    ```bash
+    export FLAGS_conv_workspace_size_limit=800 # MB
+    export FLAGS_cudnn_exhaustive_search=1
+    export FLAGS_cudnn_batchnorm_spatial_persistent=1
+    # videos数据格式
+    python3.7 -B -m paddle.distributed.launch --gpus="0,1,2,3,4,5,6,7"  --log_dir=log_timesformer  main.py  --validate -c configs/recognition/timesformer/timesformer_k400_videos.yaml
```
Comment (huangjun12) [inline docs/zh-CN/model_zoo/recognition/timesformer.md:71]: 加 --amp

File: docs/zh-CN/model_zoo/recognition/timesformer.md [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+- 由于TimeSformer模型测试模式的采样方式是速度稍慢但精度高一些的**UniformCrop**，与训练过程中验证模式采用的**RandomCrop**不同，所以训练日志中记录的验证指标`topk Acc`不代表最终的测试分数，因此在训练完成之后可以用测试模式对最好的模型进行测试获取最终的指标，命令如下：
+
+  ```bash
+  python3.7 -B -m paddle.distributed.launch --gpus="0,1,2,3,4,5,6,7"  --log_dir=log_timesformer  main.py  --test -c configs/recognition/timesformer/timesformer_k400_videos.yaml -w "output/TimeSformer/TimeSformer_best.pdparams"
+  ```
+
+
+  当测试配置采用如下参数时，在Kinetics-400的validation数据集上的测试指标如下：
+
```
Comment (huangjun12) [inline docs/zh-CN/model_zoo/recognition/timesformer.md:93]: distill这一列可以去掉，TimeSformer模型本身没有这个属性，只有优化的PP系列的才有。

File: docs/zh-CN/model_zoo/recognition/timesformer.md [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+   |      backbone      | Sampling method | distill | num_seg | target_size | Top-1 |                         checkpoints                          |
+   | :----------------: | :-------------: | :-----: | :-----: | :---------: | :---- | :----------------------------------------------------------: |
+   | Vision Transformer |   UniformCrop   |  False  |    8    |     224     | 77.08 | [TimeSformer_k400.pdparams](https://videotag.bj.bcebos.com/PaddleVideo-release2.2/TimeSformer_k400.pdparams) |
+
+
+- 测试时，TimeSformer视频采样策略为使用Linspace采样：时序上，从带采样的视频序列中均匀生成`num_seg`个稀疏采样点；空间上，对左中右或上中下3个区域采样出224尺寸的图片，一共得到3个采样区域。1个视频共采样1个clip。
```
Comment (huangjun12) [inline docs/zh-CN/model_zoo/recognition/timesformer.md:99]: 从【待】采样；
测试时，空间上，选择长边两端及中间位置（左中右 或 上中下）3个区域采样。

File: paddlevideo/loader/pipelines/augmentations.py [hunk 7]
```diff
@@ -628,3 +640,44 @@ def __call__(self, results):
 
         results['imgs'] = img_crops
         return results
+
+
+@PIPELINES.register()
+class UniformCrop:
+    """
+    Perform uniform spatial sampling on the images.
+    and then flip the cropping result to get 10 cropped images, which can make the prediction result more robust.
```
Comment (huangjun12) [inline paddlevideo/loader/pipelines/augmentations.py:649]: 注释修改下

File: paddlevideo/metrics/uniform_crop_metric.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
+# Copyright (c) 2020  PaddlePaddle Authors. All Rights Reserved.
```
Comment (huangjun12) [inline paddlevideo/metrics/uniform_crop_metric.py:1]: 代码跟center_crop_metric.py一样？ 配置文件里直接用CenterCropMetric吧， 不用在单独写一个一样的Metric类了。
上传的TenCropMetric也可以一起修一下

File: paddlevideo/tasks/train.py [hunk 9]
```diff
@@ -147,15 +174,25 @@ def train_model(cfg,
                 outputs = model(data, mode='train')
 
                 # 4.2 backward
+                if use_gradient_accumulation and i == 0:  # use GRADIENT_ACCUMULATION
+                    optimizer.clear_grad()
                 avg_loss = outputs['loss']
```
Comment (huangjun12) [inline paddlevideo/tasks/train.py:179]: 梯度累加时，loss需要除以num_iters进行缩放？

File: paddlevideo/modeling/framework/recognizers/recognizer2d.py [hunk 10]
```diff
@@ -22,18 +22,25 @@
 class Recognizer2D(BaseRecognizer):
     """2D recognizer model framework."""
     def forward_net(self, imgs):
-        #NOTE: As the num_segs is an attribute of dataset phase, and didn't pass to build_head phase, should obtain it from imgs(paddle.Tensor) now, then call self.head method.
-        num_segs = imgs.shape[1]
-        imgs = paddle.reshape_(imgs, [-1] + list(imgs.shape[2:]))
+        # NOTE: As the num_segs is an attribute of dataset phase, and didn't pass to build_head phase, should obtain it from imgs(paddle.Tensor) now, then call self.head method.
+        if self.head_name != "TimeSformerHead":
```
Comment (huangjun12) [inline paddlevideo/modeling/framework/recognizers/recognizer2d.py:26]: 重写一个recognizer_transformer.py吧  
类名用 RecognizerTransformer

**Review by huangjun12 — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (HydrogenSulfate) [reply README.md:38]: done

File: README_cn.md (hunk 2 above)
Comment (HydrogenSulfate) [reply README_cn.md:66]: done

File: docs/zh-CN/model_zoo/recognition/timesformer.md (hunk 4 above)
Comment (HydrogenSulfate) [reply docs/zh-CN/model_zoo/recognition/timesformer.md:71]: done

File: docs/zh-CN/model_zoo/recognition/timesformer.md (hunk 5 above)
Comment (HydrogenSulfate) [reply docs/zh-CN/model_zoo/recognition/timesformer.md:93]: done

File: docs/zh-CN/model_zoo/recognition/timesformer.md (hunk 6 above)
Comment (HydrogenSulfate) [reply docs/zh-CN/model_zoo/recognition/timesformer.md:99]: done

File: paddlevideo/metrics/uniform_crop_metric.py (hunk 8 above)
Comment (HydrogenSulfate) [reply paddlevideo/metrics/uniform_crop_metric.py:1]: done

File: paddlevideo/tasks/train.py (hunk 9 above)
Comment (HydrogenSulfate) [reply paddlevideo/tasks/train.py:179]: done

File: paddlevideo/modeling/framework/recognizers/recognizer2d.py (hunk 10 above)
Comment (HydrogenSulfate) [reply paddlevideo/modeling/framework/recognizers/recognizer2d.py:26]: done
